kytos/core/link.py

```diff
@@ -130,56 +131,50 @@ def available_tags(self):
         return [tag for tag in self.endpoint_a.available_tags if tag in
                 self.endpoint_b.available_tags]
 
-    def use_tag(self, tag):
-        """Remove a specific tag from available_tags if it is there.
-
-        Deprecated: use only the get_next_available_tag method.
-        """
-        if self.is_tag_available(tag):
-            self.endpoint_a.use_tag(tag)
-            self.endpoint_b.use_tag(tag)
-            return True
-        return False
-
     def is_tag_available(self, tag):
         """Check if a tag is available."""
         return (self.endpoint_a.is_tag_available(tag) and
                 self.endpoint_b.is_tag_available(tag))
 
-    def get_next_available_tag(self):
+    def get_next_available_tag(self, controller, tag_type: str = '1') -> TAG:
         """Return the next available tag if exists."""
         with self._get_available_vlans_lock:
             # Copy the available tags because in case of error
             # we will remove and add elements to the available_tags
-            available_tags_a = self.endpoint_a.available_tags.copy()
-            available_tags_b = self.endpoint_b.available_tags.copy()
-
-            for tag in available_tags_a:
-                # Tag does not exist in endpoint B. Try another tag.
-                if tag not in available_tags_b:
-                    continue
-
-                # Tag already in use. Try another tag.
-                if not self.endpoint_a.use_tag(tag):
-                    continue
-
-                # Tag already in use in B. Mark the tag as available again.
-                if not self.endpoint_b.use_tag(tag):
-                    self.endpoint_a.make_tag_available(tag)
-                    continue
-
-                # Tag used successfully by both endpoints. Returning.
-                return tag
+            available_tags_a = deepcopy(self.endpoint_a.get_available_tags())
+            available_tags_b = deepcopy(self.endpoint_b.get_available_tags())
+            intersection_tags = Interface.range_intersection(available_tags_a,
```

Comment:
How the range intersection is being computed is correct, but it's computing the entire range upfront, this could be done lazily with generators if you could refactor `Interface.range_intersection` yielding each interval, just so if already intersects in the first interval you know that you've found the next available tag. 

I'm going to also ask you to review the nested ifs below, now you're computing the intersection upfront optimally, instead of iterating on each interface, so the "tag already in use" below shouldn't happen anymore.

Reply:
Looking into this

Reply:
Got it, changed `range_intersection()` to return first tag only